lib/src/io.cpp

```diff
@@ -283,205 +202,315 @@ noexcept (false) {
-        const auto err = dlis_vrl( buffer, &len, &version );
-
-        // TODO: for now record closest to VE gets the blame
-        if (err) consistent = false;
-        if (version != 1) consistent = false;
-
-        remaining = len - DLIS_VRL_SIZE;
+        outer = inner;
     }
 }
 
-void stream::reindex( const std::vector< long long >& tells,
-                      const std::vector< int >& residuals ) noexcept (false) {
-    if (tells.empty())
-        throw std::invalid_argument( "tells must be non-empty" );
-
-    if (residuals.empty())
-        throw std::invalid_argument( "residuals must be non-empty" );
-
-    if (tells.size() != residuals.size()) {
-        const auto msg = "reindex requires tells.size() which is {}) "
-                         "== residuals.size() (which is {})"
-        ;
-        const auto str = fmt::format(msg, tells.size(), residuals.size());
-        throw std::invalid_argument(str);
+void stream::seek( std::int64_t offset ) noexcept (false) {
+    if (offset < 0) {
+        const auto msg = "expected offset (which is {}) >= 0";
+        throw std::invalid_argument(fmt::format(msg, offset));
     }
 
-    // TODO: assert all-positive etc.
-    this->tells = tells;
-    this->residuals = residuals;
+    const auto err = lfp_seek(this->protocol, offset);
+    switch (err) {
+        case LFP_OK:
+            break;
+        case LFP_INVALID_ARGS:
+        case LFP_NOTIMPLEMENTED:
+        default:
+            throw std::runtime_error(lfp_errormsg( this->protocol ));
+    }
 }
 
-void stream::close() {
-    this->fs.close();
+std::int64_t stream::read( char* dst, std::int64_t offset, int n)
+noexcept (false) {
+    if (n < 0) {
+        const auto msg = "expected n (which is {}) >= 0";
+        throw std::invalid_argument(fmt::format(msg, n));
+    }
+    this->seek( offset );
+    return this->read(dst, n);
 }
 
-void stream::read( char* dst, long long offset, int n ) {
+std::int64_t stream::read( char* dst, int n )
+noexcept (false) {
     if (n < 0) {
         const auto msg = "expected n (which is {}) >= 0";
         throw std::invalid_argument(fmt::format(msg, n));
     }
 
-    if (offset < 0) {
-        const auto msg = "expected offset (which is {}) >= 0";
-        throw std::invalid_argument(fmt::format(msg, offset));
+    std::int64_t nread = -1;
+    const auto err = lfp_readinto(this->protocol, dst, n, &nread);
+    switch (err) {
+        case LFP_OK:
+        case LFP_OKINCOMPLETE:
+        case LFP_EOF:
+            break;
+        case LFP_UNEXPECTED_EOF:
+        default:
+            throw std::runtime_error(lfp_errormsg( this->protocol ));
     }
+    return nread;
+}
+
+stream_offsets findoffsets( dl::stream& file) noexcept (false) {
+    stream_offsets ofs;
+
+    std::int64_t offset = 0;
+    char buffer[ DLIS_LRSH_SIZE ];
+
+    int len = 0;
+    while (true) {
+        file.read(buffer, offset, DLIS_LRSH_SIZE);
+        if (file.eof())
+            break;
+
```

Comment:
Again good catch! This is something we should fail on I think. If it turns out to be a real problem, then maybe consider truncation warnings rather than throws?

Reply:
Yes, we should definitely implement some mechanics on allowing user access to available data in truncated files. We have an issue for this for future.

Agree that we should fail on it now. It was a previous implementation problem that it didn't.